design/multi-client-and-multi-endpoint-in-modular.md

```diff
@@ -0,0 +1,250 @@
+1. **_Models Subpath Export_**  
+   We want to avoid exporting all our models to the top level, as this would obscure some key information about the API. Instead, we want a separate subpath for models, so that they don’t clutter the API document and can still be imported by customers if needed.
+1. **_Shared Models_**  
+   In both multi-client and multi-endpoint cases, it's possible that we can have some models are shared by both api layer sub clients, As we will have the same models in both the classical client layer and api layer, we will put those models into `src/clientA/models` and `src/clientB/models` folder, And those shared models will be in `common/models`
+1. **_Top Level `./models` and `./api` Subpath Export_**  
+   If we export both sub clients to the top level, customers will have to choose to import between the top level and subpath export from sub client. And we only need one way to import these things without there being ambiguity about which way is correct. As such, we choose to remove top level subpaths as well as the index.ts files for both models and api.
+1. **_Rest Layer Export_**  
+   We should also remove the `./rest` subpath export in Modular,  and keep the rest layer as internal in Modular to avoid  the following issues:
+    1. User experience inconsistent issues between cjs customer and esm customer as well as pure RLC customers.
+    1. To export RLC layer in Modular with  RLC as a float dependency will somehow provide extra features for Modular customers without bumping any versions.
+    1. If there’s a case where RLC layer is a breaking change and we use rename to avoid breaking in Modular layer. To export RLC layer will make it impossible to avoid that breaking in Modular.
+
+## Finalized Proposals
+
+With the above initial proposals and the above considerations, we get our finalized design for both multi-client and multi-endpoint.
+
+### Finalized Proposal in Multi-Client
+
+<!-- markdownlint-disable MD033 -->
+<table>
+  <tr>
+    <th>Subpath Exports</th>
+    <th>Source Code Structure</th>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="typescript">
+Both Exported
+@azure/foo
+
+</pre>
+<pre lang="typescript">
+ClientA
+@azure/foo/clientA
+@azure/foo/clientA/api
+@azure/foo/clientA/models
+</pre>
+<pre lang="typescript">
+ClientB
+@azure/foo/clientB
+@azure/foo/clientB/api
+@azure/foo/clientB/models
+</pre>
+</td>
+<td>
+<pre lang="typescript">
+Both Exported
+src
+src/rest
+</pre>
+<pre lang="typescript">
+ClientA
+src/clientA
+src/clientA/api
+src/clientA/models
+</pre>
+<pre lang="typescript">
+ClientB
+src/clientB
+src/clientB/api
+src/clientB/models
+</pre>
+</td>
+  </tr>
+</table>
+<!-- markdownlint-enable MD033 -->
+
+### Finalized Proposal in Multi-Endpoint
+
+<!-- markdownlint-disable MD033 -->
+<table>
+  <tr>
+    <th>Subpath Exports</th>
+    <th>Source Code Structure</th>
+  </tr>
+  <tr>
+    <td>
+      <pre lang="typescript">
+Both Exported
```

Comment:
I would be very explicit that the both exported are the classical clients. What about models?

Reply:
we will also export the models to the top level.

Reply:
updated